File: Data Scientist/Modulo III/Meus Scripts/Regressao Multipla.R
#Buscar a Correlacao entre as variaveis do dataset mtcars

cor(mtcars)

#         mpg        cyl       disp         hp
#mpg   1.0000000 -0.8521620 -0.8475514 -0.7761684
#cyl  -0.8521620  1.0000000  0.9020329  0.8324475
#disp -0.8475514  0.9020329  1.0000000  0.7909486
#hp   -0.7761684  0.8324475  0.7909486  1.0000000

#MPG sera a variavel de resposta. Inicialmente usarei DISP como variavel explanatoria

modelo = lm(mpg ~ disp, data = mtcars)
summary(modelo)

#Retorna o R Quadrado
summary(modelo)$r.squared 

#Retorna o R Ajustado
summary(modelo)$adj.r.squared

#Geracao do Grafico de dispersao dos dados do modelo e da linha e melhor ajuste
plot(mpg ~ disp, data = mtcars)
abline(modelo)
predict(modelo, data.frame(disp=200))

#Adicao de Mais variaveis explanatorias para explicacao do consumo MPG
modelo = lm(mpg ~ disp+hp+cyl+wt, data = mtcars)
summary(modelo)$r.squared
summary(modelo)$adj.r.squared
summary(modelo)

predict(modelo, data.frame(disp=200, hp=100, cyl=4))

teste <- mtcars
cor.test(mtcars$mpg, mtcars$qsec)

File: Data Scientist/Modulo III/Meus Scripts/Regressao Multipla.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sat Nov  3 15:39:32 2018

@author: victortelles
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from yellowbrick.regressor import ResidualsPlot
import statsmodels.formula.api as sm

base = pd.read_csv('Dados/mt_cars.csv')
base = base.drop('Unnamed: 0', axis = 1)

#MPG sera a variavel de resposta. Inicialmente usarei DISP como variavel explanatoria
#correlacao entre o consumo e as polegadas quadradas

X = base.iloc[:,2].values
Y = base.iloc[:,0].values

correlacao = np.corrcoef(X, Y)
correlacao
#ajuste para Linear Model - formato necessario exigido pela sklearn
X1 = X.reshape(-1, 1)

modelo = LinearRegression()
modelo.fit(X1, Y)
modelo.intercept_
modelo.coef_

previsoes = modelo.predict(X1)

#Investigar o R-Quadrado 
modelo.score(X1, Y)

#Investigar o R-Quadrado Ajustado
#SKLearn nao possue metodos para calculo do R-Quadrado Ajustado, é necessario importar outra biblioteca para isso statsmodels
modeloAjustado = sm.ols(formula = 'mpg ~ disp', data = base)
modeloTreinado = modeloAjustado.fit()

modeloTreinado.summary()

#visualizacao dos dados e da linha de melhor ajuste
plt.scatter(X1, Y)
plt.plot(X1, previsoes, color='red')

#previsoes
modelo.predict(200)

#Regressao Multipla - Adicionando mais variaveis ao Modelo
Xm = base.iloc[:,1:4].values
Ym = base.iloc[:,0].values

modeloMulti = LinearRegression()
modeloMulti.fit(Xm, Ym)

#Investigar o R-Quadrado 
modeloMulti.score(Xm, Ym)

#Investigar o R-Quadrado Ajustado
#SKLearn nao possue metodos para calculo do R-Quadrado Ajustado, é necessario importar outra biblioteca para isso statsmodels
modeloAjustadoMulti = sm.ols(formula = 'mpg ~ disp+cyl+hp', data = base)
modeloTreinadoMulti = modeloAjustadoMulti.fit()
modeloTreinadoMulti.summary()

novo = np.array([4, 200, 100])
novo = novo.reshape(1,-1)

modeloMulti.predict(novo)







